### ISSUE TITLE
403 Forbidden

### ISSUE THREAD

**Issue by aceeric** [libwww-perl/libwww-perl#412]
This is not an issue - it's a request for help. This relates to using Net::Amazon::S3 package (see https://github.com/rustyconover/net-amazon-s3/issues/118.) The aforementioned Perl package uses LWP::UserAgent to communicate with Amazon S3. There appears to be standard guidance found elsewhere on the web on how to configure LWP::UserAgent to *not* perform hostname verification of the upstream (S3) cert:
```
use Net::SSL;
$ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;
```
E.g.: https://stackoverflow.com/questions/12697450/using-lwp-with-ssl-and-client-certificates

My context is I'm using the Net::Amazon::S3 package and by inclusion the LWP::UserAgent package to access an in-house S3-compliant object store (not s3.amazon.com). I'm unable to figure out how to get these packages to successfully connect to the in house bucket. I *can* successfully connect to the in-house bucket with Python/Boto3 but I need to get it done with Perl. The LWP::UserAgent package version is 6.31 - which is what is installed when you install Net::Amazon::S3.

As stated in the ticket referenced above the response being reported by my Perl script is:
```
Invalid bucket head response: https://our-s3-endpoint.mycompany.org/test-bucket
HTTP/1.1 403 Forbidden
Date: ...
Accept-Ranges: bytes
Server: ...
WWW-Authenticate: Basic realm=...
Content-Length: 253
Content-Type: application/xml
Client-Date: ...
Client-Peer:...
Client-Response-Num: 1
Client-SSL-Cert-Issuer: <our root ca>
Client-SSL-Cert-Subject: .../CN=our-s3-endpoint.mycompany.org
Client-SSL-Cipher: ECDHE-RSA-AES128-GCM-SHA256
Client-SSL-Socket-Class: Net::SSL
Client-SSL-Warning: Peer certificate not verified
```
Any help is appreciated. Thank you.

**Comment by genio** [comment]
Do you mind joining us on IRC for some easier back and forth? https://kiwiirc.com/nextclient/#irc://irc.perl.org/#perl?nick=mc-guest-?

**Comment by aceeric** [comment]
Sure - I'm in there now. As some additional info - I noticed that if I didn't override the socket handler, that the handler was `IO::Socket::SSL` in the error message. So I put together a trivial script based on that packages guidance (https://metacpan.org/pod/IO::Socket::SSL#Basic-SSL-Client):
```
use IO::Socket::SSL;
my $cl = IO::Socket::SSL->new(
    PeerHost => "my-host",
    PeerPort => "https",
    SSL_Verify_Mode => SSL_VERIFY_NONE,
) or die ...
```
The above fails with "SSL connect attempt failed errpr 1416F086:SSL routines:tls_process_server_certificate:certifcate verify failed at ./op-soclet-ssl line 6", but the below succeeds:
```
use IO::Socket::SSL;
my $cl = IO::Socket::SSL->new(
    PeerHost => "my-host",
    PeerPort => "https",
    SSL_ca_file => "/path/to/my/ca.crt",
) or die...
```
So it is interesting that the "verify none" appears to be ignored

**Comment by Pero-Moretti** [comment]
Ok, I've been drawn in... in the code of your last comment I believe it should be SSL_verify_mode (small v, small m).

For documenting what we talked about, I asked if you could move setting $ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0 in a begin block ahead of anything else (before any use's for example). Looking down the rabbit hole, LWP::UserAgent is only looking at the environment variable once new is called, which in turn is only called when the Net::Amazon::S3 object is created, so it doesn't make sense to me that this is being ignored (over setting the environment variable before executing the script) unless the script is doing other fun stuff.

**Comment by aceeric** [comment]
Based on your comment, I got farther just setting environment variable PWDL_LWP_SSL_CA_FILE. This way, the `lwp-request` util stopped emanating the warning about peer cert not verified. I've tried that same approach with the `Net::Amazon::S3` package to no avail. So at this point, I may focus more on resolving the issues that are preventing getting `Paws` installed and set the `Net::Amazon::S3` package aside. Thanks.
